frame/contracts/src/migration.rs

```diff
@@ -266,3 +279,75 @@ mod v7 {
 		T::DbWeight::get().reads_writes(1, 2)
 	}
 }
+
+/// Update `ContractInfo` with new fields that track storage deposits.
+mod v8 {
+	use super::*;
+	use sp_io::default_child_storage as child;
+	use v6::ContractInfo as OldContractInfo;
+
+	#[derive(Encode, Decode)]
+	struct ContractInfo<T: Config> {
+		trie_id: TrieId,
+		code_hash: CodeHash<T>,
+		storage_bytes: u32,
+		storage_items: u32,
+		storage_byte_deposit: BalanceOf<T>,
+		storage_item_deposit: BalanceOf<T>,
+		storage_base_deposit: BalanceOf<T>,
+	}
+
+	#[storage_alias]
+	type ContractInfoOf<T: Config> = StorageMap<
+		Pallet<T>,
+		Twox64Concat,
+		<T as frame_system::Config>::AccountId,
+		ContractInfo<T>,
+	>;
+
+	pub fn migrate<T: Config>() -> Weight {
+		let mut weight: Weight = 0;
+
+		<ContractInfoOf<T>>::translate(|_key, old: OldContractInfo<T>| {
+			// Count storage items of this contract
+			let mut storage_bytes = 0u32;
+			let mut storage_items = 0u32;
+			let mut key = Vec::new();
+			while let Some(next) = child::next_key(&old.trie_id, &key) {
+				key = next;
+				let mut val_out = [];
+				let len = child::read(&old.trie_id, &key, &mut val_out, 0)
+					.expect("The loop conditions checks for existence of the key; qed");
+				storage_bytes.saturating_accrue(len);
+				storage_items.saturating_accrue(1);
+			}
+
+			let storage_byte_deposit =
+				T::DepositPerByte::get().saturating_mul(storage_bytes.into());
+			let storage_item_deposit =
+				T::DepositPerItem::get().saturating_mul(storage_items.into());
+			let storage_base_deposit = old
+				.storage_deposit
+				.saturating_sub(storage_byte_deposit)
+				.saturating_sub(storage_item_deposit);
+
+			// Reads: One read for each storage item plus the contract info itself.
+			// Writes: Only the new contract info.
+			weight = weight.saturating_add(
+				T::DbWeight::get().reads_writes(Weight::from(storage_items) + 1, 1),
+			);
+
+			Some(ContractInfo {
+				trie_id: old.trie_id,
+				code_hash: old.code_hash,
+				storage_bytes,
+				storage_items,
+				storage_byte_deposit,
+				storage_item_deposit,
+				storage_base_deposit,
+			})
+		});
+
+		weight
+	}
+}
```

Comment:
how do you test the migration? Normally i would expect some pre and post migration checks for try-runtime.

Reply:
I manually verified by looking at the migrated storage. Didn't know about those checks. Will add them.

Reply:
Added hooks and tested manually against some test data.